### PR TITLE
Use NIP-92 media metadata to display media in the proper orientation

### PR DESCRIPTION
**Pull request by joshuatbrown**
## Issues covered
#1172 

## Description
Gets the media metadata from the `imeta` tag, if it exists, and uses the dimensions to determine whether the gallery should be displayed in landscape or portrait.

- [x] Fix the bad code
- [x] Deduplicate code in GalleryView (inside `loadingView(_:)` where we check the metadata for orientation twice)

## How to test
There's no visible difference here. The big change is that we're using `imeta` tags from the event to determine orientation of the images/video before loading the images/video. So maybe you'll want to use [everyone's favorite proxy tool](https://proxyman.io) and the debugger to verify.

1. Navigate to Settings
2. Enable new media display
3. Navigate to Feed or Profile
4. Find some media with an `imeta` tag (`note1cj6xmz48ausu04mckl4ral8qec4j4sz7akr4uarewwchsydywraq8jwkt8`)
5. See if it's displayed correctly based on the tag

I posted some images from Damus since it includes `imeta` tags, then stepped through the debugger to be sure we're using the `imeta` tags to determine orientation instead of loading the image and checking its size to determine orientation.

## Screenshots/Video
Screenshot 2024-09-04 at 4 56 02 PM

### CODE REVIEW

File: Nos/Views/Components/RepliesLabel.swift [hunk 1]
```diff
@@ -17,7 +17,7 @@ struct RepliesLabel: View {
         self.repliesDisplayType = repliesDisplayType
 
         let noteIdentifier = note.identifier ?? ""
-        _replies = Event.replies(to: noteIdentifier)
+        _replies = FetchRequest(fetchRequest: Event.replies(to: noteIdentifier))
```
Comment (joshuatbrown) [inline Nos/Views/Components/RepliesLabel.swift:20]: This was what happened when I saw Event+CoreDataClass importing SwiftUI and removed it.

File: Nos/Models/CoreData/Event+CoreDataClass.swift
Comment (joshuatbrown) [inline Nos/Models/CoreData/Event+CoreDataClass.swift:1]: All of the changes to this file are completely outside the scope of this PR. I just wanted to try to get this file smaller as I was adding another extension for it since it's well over our file size limit in SwiftLint.

File: NosTests/IntegrationTests/EventProcessorIntegrationTests+InlineMetadata.swift [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        let squareImageMetadata = try XCTUnwrap(inlineMetadata[squareImageURL])
+        XCTAssertEqual(squareImageMetadata.dimensions, CGSize(width: 854, height: 854))
+        XCTAssertEqual(squareImageMetadata.orientation, .landscape)
+
+        let portraitImageURL = 
+            "https://image.nostr.build/2787ad495941dfb068fbff77f087513095a3f981f9697fcb9e51c052c6198090.jpg"
+        let portraitImageMetadata = try XCTUnwrap(inlineMetadata[portraitImageURL])
+        XCTAssertEqual(portraitImageMetadata.dimensions, CGSize(width: 1263, height: 3985))
+        XCTAssertEqual(portraitImageMetadata.orientation, .portrait)
+    }
```
Comment (mplorentz) [inline NosTests/IntegrationTests/EventProcessorIntegrationTests+InlineMetadata.swift:44]: It would be nice to have a test that verifies that things don't break on invalid data, like two imeta tags for the same URL. I assume in that sort of case we would just pick one arbitrarily.

**Review by mplorentz — APPROVED** [review]
This code is very nice. I left one optional suggestion but this looks good to go.

File: NosTests/IntegrationTests/EventProcessorIntegrationTests+InlineMetadata.swift (hunk 2 above)
Comment (joshuatbrown) [reply NosTests/IntegrationTests/EventProcessorIntegrationTests+InlineMetadata.swift:44]: Yes! I'll add a test for that!

File: Nos/Models/EventError.swift [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import Foundation
+
+/// Errors for an ``Event``.
+enum EventError: Error, LocalizedError {
```
Comment (martindsq) [inline Nos/Models/EventError.swift:4]: `LocalizedError` inherits from `Error` so I think we can get rid of `Error` here.

File: Nos/Models/CoreData/Event+InlineMetadata.swift [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    /// Returns an ``InlineMetadataCollection`` for the event, if any exists. Inline metadata comes from the
+    /// `imeta` tag as described in [NIP-92](https://github.com/nostr-protocol/nips/blob/master/92.md)
+    var inlineMetadata: InlineMetadataCollection? {
+        guard let tags = allTags as? [[String]] else {
+            return nil
+        }
+
+        let metadataTags = tags.filter({ $0.first == "imeta" })
+            .compactMap { InlineMetadataTag(imetaTag: $0) }
+        let collection = InlineMetadataCollection(tags: metadataTags)
```
Comment (martindsq) [inline Nos/Models/CoreData/Event+InlineMetadata.swift:13]: You can just return this

File: Nos/Models/CoreData/Event+CoreDataClass.swift [hunk 5]
```diff
@@ -1,39 +1,9 @@
 // swiftlint:disable file_length
 import secp256k1
-import Foundation
 import CoreData
-import RegexBuilder
-import SwiftUI
```
Comment (martindsq) [inline Nos/Models/CoreData/Event+CoreDataClass.swift:6]: 👏

**Review by martindsq — CHANGES_REQUESTED** [review]
I agree that the code looks nice. I'm just commenting two lines that could be improved, everything else looks great!

File: Nos/Models/CoreData/Event+InlineMetadata.swift (hunk 4 above)
Comment (joshuatbrown) [reply Nos/Models/CoreData/Event+InlineMetadata.swift:13]: 🤦

File: NosTests/IntegrationTests/EventProcessorIntegrationTests+InlineMetadata.swift (hunk 2 above)
Comment (joshuatbrown) [reply NosTests/IntegrationTests/EventProcessorIntegrationTests+InlineMetadata.swift:44]: Oh wow -- did you know it would crash if there are two `imeta` tags with the same URL? If not, I like your subtle suggestion to write a test. If so, bravo for seeing that just by looking at the code. Either way, I'm impressed. 😄

Comment (joshuatbrown) [reply NosTests/IntegrationTests/EventProcessorIntegrationTests+InlineMetadata.swift:44]: Added the test, which showed the crash, which I subsequently fixed. The code is less pretty but more stable, so I'd call it a win.

File: Nos/Models/EventError.swift (hunk 3 above)
Comment (joshuatbrown) [reply Nos/Models/EventError.swift]: Done!

File: Nos/Models/CoreData/Event+InlineMetadata.swift (hunk 4 above)
Comment (joshuatbrown) [reply Nos/Models/CoreData/Event+InlineMetadata.swift]: Done!

**Review by martindsq — APPROVED** [review]

File: NosTests/IntegrationTests/EventProcessorIntegrationTests+InlineMetadata.swift (hunk 2 above)
Comment (mplorentz) [reply NosTests/IntegrationTests/EventProcessorIntegrationTests+InlineMetadata.swift:44]: 😆 I did not know it would crash
